Gems/Atom/Utils/Code/Tests/StableDynamicStructOfArraysTests.cpp

```diff
@@ -0,0 +1,887 @@
+                TestItem& item = iterator.GetItem<TestStableDynamicStructOfArraysRows::TestItemIndex>();
+                success = success && (item.m_index == index);
+                success = success && (item.m_value == static_cast<float>(index));
+                success = success && (iterator.GetItem<TestStableDynamicStructOfArraysRows::UInt32Index>() == index);
+                index += 2;
+            }
+        }
+        EXPECT_TRUE(success);
+
+        // Remove the first half completely so there are a bunch of empty pages to skip
+        for (uint32_t i = 0; i < s_testCount / 2; ++i)
+        {
+            handles.at(i).Free();
+        }
+
+        // Now the page iterators should hit every other item after s_testCount / 2.
+        // By this passing, it provesthe first few page iterrators begin and end are equal (as they should be for empty pages).
+        success = true;
+        index = s_testCount / 2 + 1;
+        pageIterators = testArray.GetParallelRanges();
+
+        for (auto iteratorPair : pageIterators)
+        {
+            for (auto iterator = iteratorPair.first; iterator != iteratorPair.second; ++iterator)
+            {
+                TestItem& item = iterator.GetItem<TestStableDynamicStructOfArraysRows::TestItemIndex>();
+                success = success && (item.m_index == index);
+                success = success && (item.m_value == static_cast<float>(index));
+                success = success && (iterator.GetItem<TestStableDynamicStructOfArraysRows::UInt32Index>() == index);
+                index += 2;
+            }
+        }
+        EXPECT_TRUE(success);
+        handles.clear(); // cleanup remaining handles.
+    }
+
+
+
+    // Fixture for testing handles and ensuring the correct number of objects are created, modified, and/or destroyed
+    class StableDynamicStructOfArraysHandleTests
+        : public UnitTest::LeakDetectionFixture
+    {
+        friend class StableDynamicStructOfArraysOwner;
+        friend class MoveTest;
+    public:
+        void SetUp() override
+        {
+            s_testItemsConstructed = 0;
+            s_testItemsDestructed = 0;
+            s_testItemsModified = 0;
+        }
+
+        // Used to keep track of the number of times a constructor/destructor/function is called
+        // to validate that TestItems are being properly created, destroyed, and modified even when accessed via an interface
+        static int s_testItemsConstructed;
+        static int s_testItemsDestructed;
+        static int s_testItemsModified;
+    };
+
+    int StableDynamicStructOfArraysHandleTests::s_testItemsConstructed = 0;
+    int StableDynamicStructOfArraysHandleTests::s_testItemsDestructed = 0;
+    int StableDynamicStructOfArraysHandleTests::s_testItemsModified = 0;
+
+    // Class used to test that the right number of items are created, modified, and destroyed
+    // Follows similar pattern to what a FeatureProcessor might do
+    class StableDynamicStructOfArraysOwner
```

Comment:
Since the `StableDynamicStructOfArrays` was written to be used in scenarios where the performance `StableDynamicArray` was not sufficient.

This should add AZ Benchmark test that validates that this container performances better in those use cases.

Reply:
Great suggestion. I'm not getting the same improvements in the benchmark that I saw when using this in the MeshFeatureProcessor. I don't know why it worked so well on real data, possibly I had another change mixed in with the switch to the struct-of-arrays that was the true cause of the speedup.

But I have a pretty good idea why I'm not seeing an improvement in the benchmark. When passing in two structs, the data from the two structs is still lumped together in a single chunk of memory. I had mistakenly thought that the aligned_storage_t in the original StableDynamicArray was a reference to the actual data in a separate allocation from the Page itself, but now I can see that the aligned storage is part of the page itself. 

So in this case, a tuple of aligned_storage for two structs is no better than a single aligned storage for one. I think what I really need is pointers to the data from each row, plus the ability to pass in a different pool allocator for each row so that the page data for each row is closer together in memory. I will have to experiment more with the benchmark.